### ISSUE TITLE
mzn-cbc doesn't display solution as expected

### ISSUE THREAD

**Issue by skosch** [MiniZinc/libminizinc#83]
Hi,
not sure if issues against the develop branch are welcome, but in case they are ... ;)

This may be related to #76?

My relatively simple (and randomly chosen) [test case](https://gist.github.com/skosch/6022dec880eaa2de3ee7) runs into the following error:

```
▶ build/mzn-cbc --stdlib-dir share/minizinc -G linear /tmp/test.mzn   
mzn-cbc: /home/sebastian/libminizinc/include/minizinc/ast.hpp:153: MiniZinc::ASTString MiniZinc::Id::v() const: Assertion `(reinterpret_cast<ptrdiff_t>(_v_or_idn) & static_cast<ptrdiff_t>(1)) == 0' failed.
[1]    6090 abort (core dumped)  build/mzn-cbc --stdlib-dir share/minizinc -G linear /tmp/test.mzn
```

When I feed the file through `mzn2fzn` first, and then through `mzn-cbc`, the problem gets solved immediately but the output is shown as
```
▶ build/mzn-cbc --stdlib-dir share/minizinc -G linear /tmp/test.fzn 
Happiness = 121;
X_INTRODUCED_601 = 1;
X_INTRODUCED_602 = 0;
X_INTRODUCED_603 = 0;
X_INTRODUCED_604 = 0;
...
```
In other words, the reverse mapping of those `X_INTRODUCED` values to the original array elements isn't working. There is that `ozn` file that has them, but I haven't found a way to combine that with the output from `mzn-cbc`.

Thanks in advance! :)

P.S. Is an `-a` option planned, i.e. a CBC callback that prints solutions (in standard mzn output format) continuously as they come in?

**Comment by guidotack** [comment]
I've pushed a fix (hopefully) for the assertion failure. For your second case, you should be able to pipe the output from mzn-cbc through solns2out like this:

    build/mzn-cbc  --stdlib-dir share/minizinc -G linear /tmp/test.fzn | build/solns2out /tmp/test.ozn

**Comment by glebbelov** [comment]
Option -a for mzn-cbc: actually it works in cplex, gurobi and scip which are more powerful solvers. Is needed in cbc too? Might be tricky to implement

**Comment by skosch** [comment]
Thanks Guido, much appreciated!

Gleb, I've opened #85 :)

**Comment by skosch** [comment]
Sorry, have to reopen this. `solns2out` complains about not being able to find `reverse_map(int)`, which is in `linear/redefinitions.mzn`. Since it doesn't let me include other folders like the flattener does, I have to manually copy the definition over into `std/redefinitions.mzn` to make it shut up. I feel like there must be a proper way to handle this though; any pointers?

**Comment by glebbelov** [comment]
A complete example please?

**Comment by skosch** [comment]
Sure thing! Grab [this sample ozn file](https://gist.github.com/skosch/b65d8505c644692be4ce), and run `solns2mzn --stdlib-dir your_stdlib_folder sample.ozn`. I get:
```
MiniZinc: type error: no function or predicate with this signature found: `reverse_map(int)'
/tmp/sample.ozn:12688
```
However, after adding
```
function var bool: reverse_map(var int: x) = (x==1);
function bool: reverse_map(int: x) = (x==1);
```
to `your_stdlib_folder/redefinitions.mzn`, everything works a-okay (i.e. it sits there waiting for input). Thanks for looking into this!

**Comment by guidotack** [comment]
The problem is not solns2out, it’s that mzn2fzn didn’t copy the reverse_map into the ozn.  Can you please send us the mzn?

**Comment by skosch** [comment]
Sure, [here it is](https://gist.githubusercontent.com/skosch/b65d8505c644692be4ce/raw/dd78f3c5fc8e12fbd9678967711a933ae6a5d90d/sample.mzn). I have a strong suspicion that I'm doing something dumb. Do I have to add the linear redefinitions explicitly?

**Comment by glebbelov** [comment]
How do you run it? With -G linear?

**Comment by skosch** [comment]
Yes sir!

**Comment by glebbelov** [comment]
In the current develop branch, I've just tried the sequence
```
 mzn2fzn -G linear sample_sKosch.mzn
 mzn-cbc sample_sKosch.fzn |solns2out sample_sKosch.ozn
```
The .ozn contains
```
bool: X_INTRODUCED_1679 = reverse_map(X_INTRODUCED_3865);
test reverse_map(int: x) = x==1;
int: X_INTRODUCED_3865;
```
and all seems fine. Not sure about master.

**Comment by skosch** [comment]
Turns out it was my fault entirely. The reason was that I hadn't maintained the exact directory structure (`std/` adjacent to `linear/`) expected by the code. Reverting everything to factory settings fixes the problem.
